addons/mail/static/src/messaging/entity/entity/entity.js

```diff
@@ -0,0 +1,824 @@
+            }
+            const entity = new this(data);
+            state.entities[entity.localId] = entity;
+            // ensure observable version of entity is handled.
+            const proxifiedEntity = state.entities[entity.localId];
+            this.__update(proxifiedEntity, data);
+            return proxifiedEntity;
+        }
+
+        /**
+         * @static
+         */
+        static delete() {
+            for (const entity of this.all) {
+                entity.delete();
+            }
+            this.__delete(this);
+        }
+
+        /**
+         * @static
+         * @param {any} id
+         * @returns {mail.messaging.entity.Entity|undefined}
+         */
+        static fromId(id) {
+            const allEntities = this.all;
+            return allEntities.find(entity => entity.id === id);
+        }
+
+        /**
+         * @static
+         * @param {string|mail.messaging.entity.Entity|undefined} entityOrLocalId
+         * @returns {mail.messaging.entity.Entity|undefined}
+         */
+        static get(entityOrLocalId) {
+            const { state } = this.env.store;
+            if (entityOrLocalId === undefined) {
+                return undefined;
+            }
+            const entity = state.entities[
+                entityOrLocalId.isEntity
+                    ? entityOrLocalId.localId
+                    : entityOrLocalId
+            ];
+            if (!(entity instanceof this) && entity !== this) {
+                return;
+            }
+            return entity;
+        }
+
+        static link(data) {
+            this.__link(this, data);
+        }
+
+        /**
+         * @static
+         * @param {Object} data
+         * @param {any} data.id
+         * @returns {mail.messaging.entity.Entity}
+         */
+        static insert(data) {
+            let entity = this.fromId(data.id);
+            if (!entity) {
+                entity = this.create(data);
+            } else {
+                entity.update(data);
+            }
+            return entity;
+        }
+
+        /**
+         * @throws {Error} when called with non singleton classes
+         * @returns {mail.messaging.entity.Entity|undefined}
+         */
+        static get instance() {
+            if (!this.isSingleton) {
+                throw new Error(`Cannot get instance of non-singleton Entity class "${this.name}"`);
+            }
+            return this.all[0];
+        }
+
+        /**
+         * @static
+         * @returns {boolean}
+         */
+        static get isEntity() {
+            return true;
+        }
+
+        /**
+         * @static
+         * @returns {boolean}
+         */
+        static get isEntityInstance() {
+            return false;
+        }
+
+        /**
+         * @static
+         * @returns {boolean}
+         */
+        static get isEntityClass() {
+            return true;
+        }
```

Comment:
What's the point? This seems hacky, any object can think it's an entity if it has those methods.